File: packages/kits/VaultKit/rust/src/palace_payload_envelope.rs
//! palace_payload_envelope — the lossless encoding scheme for `NoteIR` fields
//! MemPalace's `add_drawer` cannot carry natively. Rust parallel of the Swift
//! `PalacePayloadEnvelope`; byte-identical envelope output for the same note.
//!
//! ## Why an envelope exists
//!
//! MemPalace's `mempalace_add_drawer` accepts exactly `wing`, `room`,
//! `content`, and the optional metadata `source_file` + `added_by`. A `NoteIR`
//! carries far more — frontmatter, wikilinks, an origin date, an attachment
//! ref, a substrate lineage UUID, KG facts, tags, a scope namespace, a kind
//! discriminator, the full path hierarchy, and the stable source key. The
//! pump's mandate is ZERO LOSS: whatever the target cannot accept as a native
//! argument is ENCODED into the payload so a re-import recovers it.
//!
//! ## The encoding (versioned, self-describing, recoverable)
//!
//! The drawer's stored `content` is:
//!
//! ```text
//! <verbatim note body>
//!
//! <!-- MOOT-ENVELOPE v1
//! { ...canonical JSON of the unmappable fields... }
//! MOOT-ENVELOPE -->
//! ```
//!
//! - The body above the marker is the note's verbatim `flattened_body`, so a
//!   MemPalace `search` (which embeds the whole content) still indexes the
//!   real prose and a human sees their note first.
//! - The marker is an HTML comment so Markdown viewers hide it; MemPalace
//!   stores content verbatim, so it survives byte-for-byte.
//! - The JSON is canonical (sorted keys, serde_json default — no slash
//!   escaping), matching Swift's `.sortedKeys`/`.withoutEscapingSlashes`, so
//!   both ports emit byte-identical envelopes.
//! - `v1` is a format version; decode refuses a version it does not know.

use crate::note_ir::{FactIR, NoteIR, OccurredAt, SourceRef, WikiLink};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Current envelope format version. Bumped only when the payload shape
/// changes; [`decode`] refuses any other version.
pub const FORMAT_VERSION: i64 = 1;

/// The opening marker prefix (carries the version digits, then a newline).
const OPEN_MARKER_PREFIX: &str = "<!-- MOOT-ENVELOPE v";
/// The closing marker.
const CLOSE_MARKER: &str = "MOOT-ENVELOPE -->";

/// The recoverable record of a `NoteIR`'s fields that MemPalace's `add_drawer`
/// native argument surface cannot carry. Serialized into the drawer content as
/// a fenced, versioned, canonical-JSON block. Field names match the Swift
/// `PalaceEnvelopePayload` Codable keys verbatim (cross-language conformance).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PalaceEnvelopePayload {
    /// `NoteIR.stable_source_key` — the idempotency key (redundant with the
    /// native `source_file` arg, carried so re-import is exact).
    #[serde(rename = "stableSourceKey")]
    pub stable_source_key: String,

    /// `NoteIR.frontmatter` — the flat provenance/anchor map.
    pub frontmatter: HashMap<String, String>,

    /// `NoteIR.links` — wikilinks (MemPalace has no wikilink concept).
    pub links: Vec<WikiLink>,

    /// `NoteIR.tags` — `#tags` (no native tag arg).
    pub tags: Vec<String>,

    /// `NoteIR.original_path` — the joined hierarchy back-compat view.
    #[serde(rename = "originalPath")]
    pub original_path: String,

    /// `NoteIR.origin_date` — when the content occurred.
    #[serde(rename = "originDate", default, skip_serializing_if = "Option::is_none")]
    pub origin_date: Option<OccurredAt>,

    /// `NoteIR.source` — an attachment ref.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source: Option<SourceRef>,

    /// `NoteIR.moot_id` — the substrate lineage UUID. Serialized uppercase to
    /// match Foundation's `UUID` (and the Swift envelope), via the shared
    /// note_ir serializer.
    #[serde(
        rename = "mootID",
        default,
        skip_serializing_if = "Option::is_none",
        serialize_with = "crate::note_ir::serialize_uuid_uppercase_pub"
    )]
    pub moot_id: Option<uuid::Uuid>,

    /// `NoteIR.facts` — KG facts (ride the envelope, never dropped).
    pub facts: Vec<FactIR>,

    /// `NoteIR.path_components` — the full ordered hierarchy.
    #[serde(rename = "pathComponents")]
    pub path_components: Vec<String>,

    /// `NoteIR.scope` — the source-tool scoping-id namespace. `BTreeMap` so
    /// serialization is deterministic (matches Swift `.sortedKeys`).
    pub scope: std::collections::BTreeMap<String, String>,

    /// `NoteIR.kind` — the entry discriminator.
    pub kind: String,
}

impl PalaceEnvelopePayload {
    /// Build the payload from a `NoteIR`, copying every field the native
    /// `add_drawer` arg surface cannot carry (plus the redundant
    /// `stable_source_key`/`path_components` for a self-sufficient record).
    pub fn from_note(note: &NoteIR) -> Self {
        Self {
            stable_source_key: note.stable_source_key.clone(),
            frontmatter: note.frontmatter.clone(),
            links: note.links.clone(),
            tags: note.tags.clone(),
            original_path: note.original_path.clone(),
            origin_date: note.origin_date.clone(),
            source: note.source.clone(),
            moot_id: note.moot_id,
            facts: note.facts.clone(),
            path_components: note.path_components.clone(),
            scope: note.scope.clone(),
            kind: note.kind.clone(),
        }
    }
}

/// Errors raised while decoding an envelope from stored drawer content.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EnvelopeDecodeError {
    /// The content carried an open marker whose version is unknown. Carries
    /// the offending version (loud failure, never a silent drop).
    UnsupportedVersion(i64),
    /// The open marker was present but the close marker was not.
    Unterminated,
    /// The bytes between the markers were not valid envelope JSON.
    MalformedJson(String),
}

impl std::fmt::Display for EnvelopeDecodeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            EnvelopeDecodeError::UnsupportedVersion(v) => {
                write!(f, "unsupported envelope version {v}")
            }
            EnvelopeDecodeError::Unterminated => write!(f, "unterminated envelope"),
            EnvelopeDecodeError::MalformedJson(m) => write!(f, "malformed envelope JSON: {m}"),
        }
    }
}

impl std::error::Error for EnvelopeDecodeError {}

/// One decoded drawer: the verbatim body with the envelope stripped, plus the
/// recovered payload (`None` when the content carried no envelope).
#[derive(Debug, Clone, PartialEq)]
pub struct Decoded {
    /// The note prose with the fenced envelope removed and surrounding
    /// whitespace trimmed.
    pub body: String,
    /// The recovered unmappable-field record, or `None` when absent.
    pub payload: Option<PalaceEnvelopePayload>,
}

/// Fold a note's body and its unmappable fields into the drawer content
/// MemPalace will store: the verbatim body, a blank line, then the fenced
/// envelope. Inverse of [`decode`].
pub fn encode(body: &str, payload: &PalaceEnvelopePayload) -> Result<String, EnvelopeDecodeError> {
    let json = canonical_json(payload)?;
    Ok(format!(
        "{body}\n\n{OPEN_MARKER_PREFIX}{FORMAT_VERSION}\n{json}\n{CLOSE_MARKER}"
    ))
}

/// One decoded four-noun envelope: the verbatim body with the envelope
/// stripped, plus the recovered per-noun field map (empty when the content
/// carried no envelope). Mirrors Swift `PalacePayloadEnvelope.DecodedFields`.
#[derive(Debug, Clone, PartialEq)]
pub struct DecodedFields {
    /// The native text with the fenced envelope removed and surrounding
    /// whitespace trimmed.
    pub body: String,
    /// The recovered per-noun envelope-field map (empty when none present).
    pub fields: std::collections::BTreeMap<String, serde_json::Value>,
}

/// Fold a `PalaceItem`'s body and its per-noun envelope-field map into the
/// string MemPalace stores for whichever native arg carries the noun's text.
/// The ONE canonical envelope for ALL four nouns — the same versioned
/// `MOOT-ENVELOPE v1` marker as [`encode`]. When `fields` is empty the body is
/// returned UNCHANGED (no empty envelope), mirroring the per-noun mapping's
/// "ride only when there is something to carry" rule and the Swift
/// `encodeFields`. Bytes match Swift for identical input.
pub fn encode_fields(
    body: &str,
    fields: &std::collections::BTreeMap<String, serde_json::Value>,
) -> Result<String, EnvelopeDecodeError> {
    if fields.is_empty() {
        return Ok(body.to_owned());
    }
    let json = canonical_fields_json(fields)?;
    Ok(format!(
        "{body}\n\n{OPEN_MARKER_PREFIX}{FORMAT_VERSION}\n{json}\n{CLOSE_MARKER}"
    ))
}

/// Split a stored four-noun string back into its body and recovered field map.
/// The inverse of [`encode_fields`]. Content with no open marker decodes as
/// `{ body: content, fields: {} }`. Mirrors the Swift `decodeFields`.
pub fn decode_fields(content: &str) -> Result<DecodedFields, EnvelopeDecodeError> {
    let open_idx = match content.find(OPEN_MARKER_PREFIX) {
        Some(i) => i,
        None => {
            return Ok(DecodedFields {
                body: content.to_owned(),
                fields: std::collections::BTreeMap::new(),
            })
        }
    };
    let after_prefix = &content[open_idx + OPEN_MARKER_PREFIX.len()..];
    let line_end = after_prefix
        .find('\n')
        .ok_or(EnvelopeDecodeError::Unterminated)?;
    let version_token = after_prefix[..line_end].trim();
    let version: i64 = version_token.parse().map_err(|_| {
        EnvelopeDecodeError::MalformedJson(format!(
            "envelope version token '{version_token}' is not an integer"
        ))
    })?;
    if version != FORMAT_VERSION {
        return Err(EnvelopeDecodeError::UnsupportedVersion(version));
    }
    let json_start = open_idx + OPEN_MARKER_PREFIX.len() + line_end + 1;
    let close_rel = content[json_start..]
        .find(CLOSE_MARKER)
        .ok_or(EnvelopeDecodeError::Unterminated)?;
    let json_slice = content[json_start..json_start + close_rel].trim();
    let fields: std::collections::BTreeMap<String, serde_json::Value> =
        serde_json::from_str(json_slice)
            .map_err(|e| EnvelopeDecodeError::MalformedJson(e.to_string()))?;
    let body = content[..open_idx].trim().to_owned();
    Ok(DecodedFields { body, fields })
}

/// Encode a per-noun envelope-field map as canonical JSON: sorted keys (the
/// `BTreeMap` is already ordered, and serde_json re-emits in that order), no
/// slash escaping — matching Swift `.sortedKeys`/`.withoutEscapingSlashes`, so
/// both ports emit byte-identical four-noun envelopes.
fn canonical_fields_json(
    fields: &std::collections::BTreeMap<String, serde_json::Value>,
) -> Result<String, EnvelopeDecodeError> {
    // Round-trip through a Value so nested object keys are also BTree-sorted,
    // matching the typed-payload canonical_json path and the Swift output.
    let value = serde_json::to_value(fields)
        .map_err(|e| EnvelopeDecodeError::MalformedJson(e.to_string()))?;
    serde_json::to_string(&value).map_err(|e| EnvelopeDecodeError::MalformedJson(e.to_string()))
}

/// Split stored drawer content back into its body and recovered payload.
/// Content with no open marker decodes as `{ body: content, payload: None }`
/// — a foreign drawer is read as plain prose, never an error.
pub fn decode(content: &str) -> Result<Decoded, EnvelopeDecodeError> {
    let open_idx = match content.find(OPEN_MARKER_PREFIX) {
        Some(i) => i,
        None => {
            return Ok(Decoded {
                body: content.to_owned(),
                payload: None,
            })
        }
    };
    // Version digits run from after the prefix to the end of the marker line.
    let after_prefix = &content[open_idx + OPEN_MARKER_PREFIX.len()..];
    let line_end = after_prefix
        .find('\n')
        .ok_or(EnvelopeDecodeError::Unterminated)?;
    let version_token = after_prefix[..line_end].trim();
    let version: i64 = version_token
        .parse()
        .map_err(|_| EnvelopeDecodeError::MalformedJson(format!(
            "envelope version token '{version_token}' is not an integer"
        )))?;
    if version != FORMAT_VERSION {
        return Err(EnvelopeDecodeError::UnsupportedVersion(version));
    }
    // JSON runs from after the marker line to before the close marker.
    let json_start = open_idx + OPEN_MARKER_PREFIX.len() + line_end + 1;
    let close_rel = content[json_start..]
        .find(CLOSE_MARKER)
        .ok_or(EnvelopeDecodeError::Unterminated)?;
    let json_slice = content[json_start..json_start + close_rel].trim();
    let payload: PalaceEnvelopePayload = serde_json::from_str(json_slice)
        .map_err(|e| EnvelopeDecodeError::MalformedJson(e.to_string()))?;
    let body = content[..open_idx].trim().to_owned();
    Ok(Decoded {
        body,
        payload: Some(payload),
    })
}

/// Reconstruct a full `NoteIR` from a fetched drawer's content. When no
/// envelope is present, a minimal note is built from the prose with
/// `stable_source_key = fallback_key`.
pub fn reconstruct_note(content: &str, fallback_key: &str) -> Result<NoteIR, EnvelopeDecodeError> {
    use crate::note_ir::Block;
    let decoded = decode(content)?;
    match decoded.payload {
        None => Ok(NoteIR::new(
            fallback_key,
            vec![Block::markdown(decoded.body)],
            HashMap::new(),
            Vec::new(),
            Vec::new(),
            "",
            None,
            None,
        )),
        Some(p) => {
            let mut note = NoteIR::with_moot_id(
                p.stable_source_key,
                vec![Block::markdown(decoded.body)],
                p.frontmatter,
                p.links,
                p.tags,
                p.original_path,
                p.origin_date,
                p.source,
                p.moot_id,
            );
            note.facts = p.facts;
            note.path_components = p.path_components;
            note.scope = p.scope;
            note.kind = p.kind;
            Ok(note)
        }
    }
}

/// Encode the payload as canonical JSON: serde_json serializes struct fields
/// in declaration order, but the cross-language anchor requires KEY-SORTED
/// output (matching Swift `.sortedKeys`). We therefore round-trip through a
/// `serde_json::Value` and re-serialize, whose object maps are BTree-backed
/// and thus sorted — and serde_json never escapes '/'.
fn canonical_json(payload: &PalaceEnvelopePayload) -> Result<String, EnvelopeDecodeError> {
    let value: serde_json::Value = serde_json::to_value(payload)
        .map_err(|e| EnvelopeDecodeError::MalformedJson(e.to_string()))?;
    serde_json::to_string(&value).map_err(|e| EnvelopeDecodeError::MalformedJson(e.to_string()))
}

File: packages/kits/VaultKit/Sources/VaultKit/PalacePayloadEnvelope.swift
import Foundation

// PalacePayloadEnvelope.swift — the lossless encoding scheme for NoteIR
// fields MemPalace's `add_drawer` cannot carry natively.
//
// ## Why an envelope exists
//
// MemPalace's `mempalace_add_drawer` accepts exactly: `wing`, `room`,
// `content`, and the optional metadata `source_file` + `added_by`. A
// `NoteIR` carries far more than that — frontmatter, wikilinks, an origin
// date, an attachment ref, a substrate lineage UUID, KG facts, tags, a
// scope namespace, a kind discriminator, the full path hierarchy, and the
// stable source key. The pump's mandate is ZERO LOSS: whatever the target
// cannot accept as a native argument is ENCODED into the payload so a
// re-import recovers it.
//
// The mappable fields (`wing`/`room` derived from `pathComponents`, the
// note body as `content`, the stable key as `source_file`) ride the native
// args; PalacePumpMapping handles those. THIS file owns the rest: the
// fields the native arg surface cannot express are serialized into a single
// fenced block appended to the drawer content.
//
// ## The encoding (versioned, self-describing, recoverable)
//
// The drawer's stored `content` is:
//
//     <verbatim note body>
//     \n
//     <!-- MOOT-ENVELOPE v1
//     { ...canonical JSON of the unmappable fields... }
//     MOOT-ENVELOPE -->
//
// - The body above the marker is the note's verbatim `flattenedBody`, so a
//   MemPalace `search` (which embeds the whole content) still indexes the
//   real prose, and a human reading the drawer sees their note first.
// - The marker is an HTML comment so it renders invisibly in any Markdown
//   viewer (MemPalace stores Markdown-ish prose) yet survives verbatim
//   storage byte-for-byte — MemPalace stores content unmodified (verified:
//   get_drawer returns `content` exactly as written).
// - The JSON is canonical (sorted keys, no slash escaping) so the Swift and
//   Rust ports emit byte-identical envelopes for the same `NoteIR` — the
//   cross-language conformance anchor for this codec.
// - `v1` is a format version: a future field addition bumps it, and decode
//   refuses a version it does not understand rather than silently dropping.
//
// ## What rides the envelope (the unmappable set)
//
// Enumerated against the `add_drawer` native arg surface. Every `NoteIR`
// field that is NOT (`body` → content, `pathComponents` → wing/room,
// `stableSourceKey` → source_file) is carried here:
//
//   frontmatter, links, tags, originDate, source, mootID, facts, scope,
//   kind, originalPath, stableSourceKey, pathComponents.
//
// `stableSourceKey` and `pathComponents` are encoded REDUNDANTLY (they also
// ride native args) so a re-import that reads ONLY the envelope reconstructs
// the full `NoteIR` without having to reverse-engineer wing/room back into a
// path or trust the lossy `source_file` round-trip. Redundancy is cheap and
// makes the envelope a complete, self-sufficient record.

/// The recoverable record of a `NoteIR`'s fields that MemPalace's
/// `add_drawer` native argument surface cannot carry. Serialized into the
/// drawer content as a fenced, versioned, canonical-JSON block by
/// ``PalacePayloadEnvelope`` so a re-import loses nothing.
///
/// Every field mirrors the `NoteIR` field of the same name. `Codable` with
/// canonical output; `Equatable` so the round-trip test asserts
/// `decode(encode(x)) == x`.
public struct PalaceEnvelopePayload: Codable, Sendable, Equatable {

    /// `NoteIR.stableSourceKey` — carried so a re-import recovers the exact
    /// idempotency key (redundant with the native `source_file` arg).
    public var stableSourceKey: String

    /// `NoteIR.frontmatter` — the flat provenance/anchor map. No native arg.
    public var frontmatter: [String: String]

    /// `NoteIR.links` — wikilinks. MemPalace has no wikilink concept.
    public var links: [WikiLink]

    /// `NoteIR.tags` — `#tags`. `add_drawer` carries no tag arg.
    public var tags: [String]

    /// `NoteIR.originalPath` — the joined hierarchy back-compat view.
    public var originalPath: String

    /// `NoteIR.originDate` — when the content occurred. No native arg.
    public var originDate: OccurredAt?

    /// `NoteIR.source` — an attachment ref. No native arg.
    public var source: SourceRef?

    /// `NoteIR.mootID` — the substrate lineage UUID. No native arg.
    public var mootID: UUID?

    /// `NoteIR.facts` — KG facts. MemPalace has its own KG, but the pump
    /// writes drawers; the facts ride the envelope so a re-import recovers
    /// them losslessly rather than dropping the graph layer.
    public var facts: [FactIR]

    /// `NoteIR.pathComponents` — the full ordered hierarchy (redundant with
    /// the native wing/room mapping, carried so re-import is exact).
    public var pathComponents: [String]

    /// `NoteIR.scope` — the source-tool scoping-id namespace. No native arg.
    public var scope: [String: String]

    /// `NoteIR.kind` — the entry discriminator (`note`/`fact`/`journal`).
    public var kind: String

    public init(
        stableSourceKey: String,
        frontmatter: [String: String],
        links: [WikiLink],
        tags: [String],
        originalPath: String,
        originDate: OccurredAt?,
        source: SourceRef?,
        mootID: UUID?,
        facts: [FactIR],
        pathComponents: [String],
        scope: [String: String],
        kind: String
    ) {
        self.stableSourceKey = stableSourceKey
        self.frontmatter = frontmatter
        self.links = links
        self.tags = tags
        self.originalPath = originalPath
        self.originDate = originDate
        self.source = source
        self.mootID = mootID
        self.facts = facts
        self.pathComponents = pathComponents
        self.scope = scope
        self.kind = kind
    }

    /// Build the payload from a `NoteIR`, copying every field the native
    /// `add_drawer` arg surface cannot carry (plus the redundant
    /// `stableSourceKey`/`pathComponents` for a self-sufficient record).
    public init(from note: NoteIR) {
        self.init(
            stableSourceKey: note.stableSourceKey,
            frontmatter: note.frontmatter,
            links: note.links,
            tags: note.tags,
            originalPath: note.originalPath,
            originDate: note.originDate,
            source: note.source,
            mootID: note.mootID,
            facts: note.facts,
            pathComponents: note.pathComponents,
            scope: note.scope,
            kind: note.kind
        )
    }
}

/// The codec that folds a ``PalaceEnvelopePayload`` into (and back out of) a
/// MemPalace drawer's `content`, and reconstructs a full `NoteIR` from a
/// fetched drawer. Pure value transforms — no I/O, deterministic, and
/// byte-identical across the Swift and Rust ports.
public enum PalacePayloadEnvelope {

    /// Current envelope format version. Bumped only when the payload shape
    /// changes; ``decode(content:)`` refuses any other version.
    public static let formatVersion = 1

    /// The opening marker line. Carries the version so a decoder validates
    /// before parsing. An HTML comment so Markdown viewers hide it.
    static let openMarkerPrefix = "<!-- MOOT-ENVELOPE v"

    /// The closing marker line.
    static let closeMarker = "MOOT-ENVELOPE -->"

    /// Errors raised while decoding an envelope from stored drawer content.
    public enum DecodeError: Error, Sendable, Equatable {
        /// The content carried an open marker whose version is not one this
        /// build understands. Carries the offending version (loud failure,
        /// never a silent drop) — symmetry with `CorpusDocument` strictness.
        case unsupportedVersion(Int)
        /// The open marker was present but the matching close marker was not,
        /// so the envelope bytes are truncated/corrupt.
        case unterminated
        /// The bytes between the markers were not valid envelope JSON.
        case malformedJSON(String)
    }

    // MARK: - Encode

    /// Fold a note's body and its unmappable fields into the drawer content
    /// MemPalace will store: the verbatim body, then the fenced envelope.
    ///
    /// When the payload is entirely default (a plain note with nothing the
    /// native args dropped) the envelope is STILL emitted — `stableSourceKey`
    /// is always non-default, so a re-import always has the idempotency key.
    /// The body is emitted verbatim so the searchable prose is unchanged.
    ///
    /// - Parameters:
    ///   - body: the note's `flattenedBody` (the verbatim prose).
    ///   - payload: the unmappable-field record to fold in.
    /// - Returns: the `content` string to send as the `add_drawer` content
    ///   argument.
    public static func encode(body: String, payload: PalaceEnvelopePayload) throws -> String {
        let json = try canonicalJSON(payload)
        // Body, blank line, then the fenced block. A trailing newline after
        // the close marker is omitted so re-encoding is byte-stable.
        return "\(body)\n\n\(openMarkerPrefix)\(formatVersion)\n\(json)\n\(closeMarker)"
    }

    // MARK: - Generic four-noun envelope (drawer / tunnel / KG fact / diary)

    /// Fold a `PalaceItem`'s body and its per-noun envelope-field map into the
    /// string MemPalace stores for whichever native arg carries the noun's text
    /// (a drawer's/diary's `content`/`entry`, a tunnel's `label`, a fact's
    /// `source_closet`).
    ///
    /// This is the ONE canonical envelope for ALL four nouns — the same
    /// versioned `MOOT-ENVELOPE v1` marker as the `NoteIR` drawer envelope
    /// above. The payload is the field map, serialized as canonical JSON
    /// (sorted keys, no slash escaping) so the Swift and Rust ports emit
    /// byte-identical bytes.
    ///
    /// When `fields` is empty the body is returned UNCHANGED (no empty
    /// envelope): a fact with no extra metadata sends a clean `source_closet`,
    /// and a plain item stays plain. This mirrors the per-noun mapping's
    /// "ride only when there is something to carry" rule.
    ///
    /// - Parameters:
    ///   - body: the noun's native text (drawer content, diary entry, tunnel
    ///     label, or "" for a fact whose envelope rides `source_closet`).
    ///   - fields: the per-noun envelope-field map (every field with no native
    ///     MemPalace home).
    /// - Returns: `body` when `fields` is empty; otherwise body, a blank line,
    ///   then the fenced versioned envelope.
    public static func encodeFields(body: String, fields: [String: PalaceJSONValue]) throws -> String {
        guard !fields.isEmpty else { return body }
        let json = try canonicalFieldsJSON(fields)
        return "\(body)\n\n\(openMarkerPrefix)\(formatVersion)\n\(json)\n\(closeMarker)"
    }

    /// One decoded four-noun envelope: the verbatim body with the envelope
    /// stripped, plus the recovered field map (empty when the content carried
    /// no envelope).
    public struct DecodedFields: Sendable, Equatable {
        /// The native text with the fenced envelope removed and surrounding
        /// whitespace trimmed.
        public let body: String
        /// The recovered per-noun envelope-field map (empty when no envelope
        /// was present).
        public let fields: [String: PalaceJSONValue]
    }

    /// Split a stored four-noun string back into its body and recovered field
    /// map. The inverse of ``encodeFields(body:fields:)``. Content with no open
    /// marker decodes as `DecodedFields(body: content, fields: [:])` — a
    /// foreign value is read as plain text, never an error.
    ///
    /// - Throws: ``DecodeError`` when an envelope is present but its version is
    ///   unsupported, it is unterminated, or its JSON is malformed.
    public static func decodeFields(content: String) throws -> DecodedFields {
        guard let openRange = content.range(of: openMarkerPrefix) else {
            return DecodedFields(body: content, fields: [:])
        }
        let afterPrefix = content[openRange.upperBound...]
        guard let lineEnd = afterPrefix.firstIndex(of: "\n") else {
            throw DecodeError.unterminated
        }
        let versionToken = afterPrefix[afterPrefix.startIndex..<lineEnd]
            .trimmingCharacters(in: .whitespaces)
        guard let version = Int(versionToken) else {
            throw DecodeError.malformedJSON("envelope version token '\(versionToken)' is not an integer")
        }
        guard version == formatVersion else {
            throw DecodeError.unsupportedVersion(version)
        }
        let jsonStart = afterPrefix.index(after: lineEnd)
        guard let closeRange = content.range(of: closeMarker, range: jsonStart..<content.endIndex) else {
            throw DecodeError.unterminated
        }
        let jsonSlice = content[jsonStart..<closeRange.lowerBound]
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let fields: [String: PalaceJSONValue]
        do {
            fields = try JSONDecoder().decode(
                [String: PalaceJSONValue].self,
                from: Data(jsonSlice.utf8))
        } catch {
            throw DecodeError.malformedJSON(String(describing: error))
        }
        let body = String(content[content.startIndex..<openRange.lowerBound])
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return DecodedFields(body: body, fields: fields)
    }

    // MARK: - Decode

    /// One decoded drawer: the verbatim body with the envelope stripped, plus
    /// the recovered payload (nil when the content carried no envelope, e.g.
    /// a drawer written by something other than this pump).
    public struct Decoded: Sendable, Equatable {
        /// The note prose with the fenced envelope removed and trailing
        /// envelope-separator whitespace trimmed.
        public let body: String
        /// The recovered unmappable-field record, or nil when no envelope was
        /// present in the content.
        public let payload: PalaceEnvelopePayload?
    }

    /// Split stored drawer content back into its body and its recovered
    /// payload. The inverse of ``encode(body:payload:)``:
    /// `decode(encode(body, p)).body == body` and `.payload == p`.
    ///
    /// Content with no open marker decodes as `Decoded(body: content,
    /// payload: nil)` — a foreign drawer is read as plain prose, never an
    /// error.
    ///
    /// - Throws: ``DecodeError`` when an envelope is present but its version
    ///   is unsupported, it is unterminated, or its JSON is malformed.
    public static func decode(content: String) throws -> Decoded {
        guard let openRange = content.range(of: openMarkerPrefix) else {
            return Decoded(body: content, payload: nil)
        }
        // Parse the version digits immediately after the prefix up to the
        // newline that ends the open-marker line.
        let afterPrefix = content[openRange.upperBound...]
        guard let lineEnd = afterPrefix.firstIndex(of: "\n") else {
            throw DecodeError.unterminated
        }
        let versionToken = afterPrefix[afterPrefix.startIndex..<lineEnd]
            .trimmingCharacters(in: .whitespaces)
        guard let version = Int(versionToken) else {
            throw DecodeError.malformedJSON("envelope version token '\(versionToken)' is not an integer")
        }
        guard version == formatVersion else {
            throw DecodeError.unsupportedVersion(version)
        }
        // The JSON runs from just after the open-marker line to just before
        // the close marker.
        let jsonStart = afterPrefix.index(after: lineEnd)
        guard let closeRange = content.range(of: closeMarker, range: jsonStart..<content.endIndex) else {
            throw DecodeError.unterminated
        }
        let jsonSlice = content[jsonStart..<closeRange.lowerBound]
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let payload: PalaceEnvelopePayload
        do {
            payload = try JSONDecoder().decode(
                PalaceEnvelopePayload.self,
                from: Data(jsonSlice.utf8))
        } catch {
            throw DecodeError.malformedJSON(String(describing: error))
        }
        // Body is everything before the open marker, trimmed of the blank
        // separator line(s) encode inserted.
        let body = String(content[content.startIndex..<openRange.lowerBound])
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return Decoded(body: body, payload: payload)
    }

    /// Reconstruct a full `NoteIR` from a fetched drawer's content. Decodes
    /// the envelope and rebuilds the note from the recovered payload, with
    /// the body above the marker as the single markdown block. When the
    /// content carries no envelope (a foreign drawer), a minimal `NoteIR` is
    /// built from the prose alone with `stableSourceKey = fallbackKey`.
    ///
    /// - Parameters:
    ///   - content: the drawer content as returned by `get_drawer`.
    ///   - fallbackKey: the stable key to use when no envelope is present
    ///     (e.g. the drawer id) so the reconstructed note still has identity.
    /// - Returns: the reconstructed `NoteIR`.
    public static func reconstructNote(content: String, fallbackKey: String) throws -> NoteIR {
        let decoded = try decode(content: content)
        guard let payload = decoded.payload else {
            return NoteIR(
                stableSourceKey: fallbackKey,
                body: [Block(kind: "markdown", text: decoded.body)]
            )
        }
        return NoteIR(
            stableSourceKey: payload.stableSourceKey,
            body: [Block(kind: "markdown", text: decoded.body)],
            frontmatter: payload.frontmatter,
            links: payload.links,
            tags: payload.tags,
            originalPath: payload.originalPath,
            originDate: payload.originDate,
            source: payload.source,
            mootID: payload.mootID,
            facts: payload.facts,
            pathComponents: payload.pathComponents,
            scope: payload.scope,
            kind: payload.kind
        )
    }

    // MARK: - Canonical JSON

    /// Encode the payload as canonical JSON: sorted keys, no slash escaping —
    /// the same conventions as `CorpusDocument.canonicalJSON` and the Rust
    /// `serde_json` default, so both ports emit byte-identical envelopes.
    static func canonicalJSON(_ payload: PalaceEnvelopePayload) throws -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        let data = try encoder.encode(payload)
        return String(decoding: data, as: UTF8.self)
    }

    /// Encode a per-noun envelope-field map as canonical JSON: sorted keys, no
    /// slash escaping — the same conventions as ``canonicalJSON(_:)`` and the
    /// Rust `serde_json` default, so both ports emit byte-identical four-noun
    /// envelopes for the same field map.
    static func canonicalFieldsJSON(_ fields: [String: PalaceJSONValue]) throws -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        let data = try encoder.encode(fields)
        return String(decoding: data, as: UTF8.self)
    }
}
